src/Sarif.Multitool/Rules/SARIF2006.UrisShouldBeReachable.cs

```diff
@@ -0,0 +1,105 @@
+        public override string Id => RuleId.UrisShouldBeReachable;
+
+        /// <summary>
+        /// Placeholder
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2006_UrisShouldBeReachable_FullDescription_Text };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2006_UrisShouldBeReachable_Warning_Default_Text)
+        };
+
+        public override FailureLevel DefaultLevel => FailureLevel.Warning;
+
+        private static readonly HttpClient s_httpClient = new HttpClient();
+        private static readonly Dictionary<string, bool> s_checkedUris = new Dictionary<string, bool>();
+
+        protected override void Analyze(SarifLog log, string logPointer)
+        {
+            AnalyzeUri(log.SchemaUri, logPointer.AtProperty(SarifPropertyName.Schema));
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            if (result.WorkItemUris != null)
+            {
+                Uri[] workItemUris = result.WorkItemUris.ToArray();
+                string workItemUrisPointer = resultPointer.AtProperty(SarifPropertyName.WorkItemUris);
+
+                for (int i = 0; i < workItemUris.Length; ++i)
+                {
+                    AnalyzeUri(workItemUris[i], workItemUrisPointer.AtIndex(i));
+                }
+            }
+        }
+
+        protected override void Analyze(ReportingDescriptor reportingDescriptor, string reportingDescriptorPointer)
+        {
+            AnalyzeUri(reportingDescriptor.HelpUri, reportingDescriptorPointer.AtProperty(SarifPropertyName.HelpUri));
+        }
+
+        protected override void Analyze(ToolComponent toolComponent, string toolComponentPointer)
+        {
+            AnalyzeUri(toolComponent.DownloadUri, toolComponentPointer.AtProperty(SarifPropertyName.DownloadUri));
+        }
+
+        protected override void Analyze(VersionControlDetails versionControlDetails, string versionControlDetailsPointer)
+        {
+            AnalyzeUri(versionControlDetails.RepositoryUri, versionControlDetailsPointer.AtProperty(SarifPropertyName.RepositoryUri));
+        }
+
+        private void AnalyzeUri(Uri uri, string pointer)
+        {
+            AnalyzeUri(uri?.OriginalString, pointer);
+        }
+
+        private void AnalyzeUri(string uriString, string pointer)
+        {
+            // If it's not a well-formed URI of _any_ kind, then don't bother triggering this rule.
+            // Rule SARIF1003, UrisMustBeValid, will catch it.
+            // Check for well-formedness first, before attempting to create a Uri object, to
+            // avoid having to do a try/catch. Unfortunately Uri.TryCreate will return true
+            // even for a malformed URI string.
+            if (uriString != null && Uri.IsWellFormedUriString(uriString, UriKind.RelativeOrAbsolute))
```

Comment:
You can simplify this logic because if it's not absolute, SARIF1005.UriMustBeAbsolute will catch it. So:

```
        private void AnalyzeUri(string uriString, string pointer)
        {
            // If it's not a well-formed URI, or if it's not absolute, then don't bother triggering this rule.
            // If it's not well-formed, SARIF1003.UrisMustBeValid will catch it, and if it's not absolute,
            // SARIF1005.UriMustBeAbsolute will catch it.
            //
            // Check for well-formedness first, before attempting to create a Uri object, to
            // avoid having to do a try/catch. Unfortunately Uri.TryCreate will return true
            // even for a malformed URI string.
            if (uriString != null && Uri.IsWellFormedUriString(uriString, UriKind.Absolute))
            {
                // Ok, it's a well-formed absolute URI. If it's not reachable, _now_ we can report it.
                Uri uri = new Uri(uriString, UriKind.Absolute);
                if (!IsUriReachable(uri.AbsoluteUri))
                {
                    // Placeholder
                    LogResult(pointer, nameof(RuleResources.SARIF1005_UriMustBeAbsolute_Error_Default_Text), uriString);
                }
            }
        }
```

Reply:
If i do that and the uri is not absolute, we will have a null exception. It's why I added. Unless I use the uriString

---
In reply to: [446547297](https://github.com/microsoft/sarif-sdk/pull/1942#discussion_r446547297) [](ancestors = 446547297)

Reply:
As we discussed: I now use `UriKind.Absolute` in the call to `IsWellFormedUriString` so it should work.

---
In reply to: [446547509](https://github.com/microsoft/sarif-sdk/pull/1942#discussion_r446547509) [](ancestors = 446547509,446547297)